Mathlib/RingTheory/Binomial.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2023 Scott Carnahan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Carnahan
+-/
+
+import Mathlib.Data.Rat.NNRat
+import Mathlib.RingTheory.Polynomial.Pochhammer
+
+/-!
+# Binomial semirings and binomial rings
+
+In this file we introduce binomial semirings and binomial rings, together with related functions
+like binomial coefficients.
+
+According to our main reference [elliott2006binomial] (which lists many equivalent conditions), a
+binomial ring is a torsion-free commutative ring R such that for any `x ∈ R` and any `k ∈ ℕ`, the
+product `x(x-1)⋯(x-k+1)` is divisible by `k!`.  The torsion-free condition lets us divide by `k!`
+unambiguously, so we get uniquely defined binomial coefficients.
+
+The defining condition doesn't require commutativity, and we get a theory with essentially the same
+power by replacing subtraction with addition.  Thus, we consider a semiring R in which
+multiplication by factorials is injective, and demand that for all `x ∈ R` and any `k ∈ ℕ`,
+`x(x+1)⋯(x+(k-1))` is divisible by `k!`.
+The quotient is called `multichoose r k`, following the convention given for natural numbers.
+
+## References
+* [J. Elliott, *Binomial rings, integer-valued polynomials, and λ-rings*][elliott2006binomial]
+
+-/
+
+universe u
+
+variable {R: Type u}
+
+section Semiring
+
+variable [Semiring R]
+
+namespace Ring
+
+/-- `ascPochEval` directly evaluates the `n`-th ascending Pochhammer polynomial at an element `r`.-/
+def ascPochEval (r : R) : ℕ → R
```

Comment:
Can we give this a more distinctive name, like `risingFactorial` or `ascendingFactorial` or something?

Reply:
`Nat.ascFactorial` is similar, but is defined so that `Nat.ascFactorial n r = ascPochEval (n + 1) r`.  I didn't want to introduce any off-by-one confusion.

Reply:
I'm still not sure about this, having two very similar ways of expressing the same thing is a recipe for additional work down the line. I'd really prefer it if there was a way to unify them. It seems to me that the definition of `ascFactorial` could be nonstandard, seeing as it doesn't match Wikipedia for one. It would be good to understand why it was defined in this way in mathlib in the first place and if possible change that rather than reimplementing it

Reply:
I see there is even a big note at the top of that file saying * `Nat.ascFactorial`: The ascending factorial. Note that it runs from `n + 1` to `n + k`
  and *not* from `n` to `n + k - 1`. We might want to change that in the future. So I think that we should indeed look seriously into changing it

Reply:
I changed it to `ascFactorial`.  In fact, I think it is worth changing the existing def of `Nat.ascFactorial` since it goes against essentially all of the existing literature, and also doesn't generalize well.

Reply:
Sorry, I didn't reload the page before writing my comment, so our messages crossed.  I made a separate branch where `Nat.ascFactorial` is fixed (and I am nailing down the proofs that got broken), but now I am thinking perhaps Nat.ascFactorial should just be eliminated in favor of the general semiring ascFactorial.  Should I make a separate RingTheory/ascFactorial.lean file with few dependencies carrying just the definition and a few properties?